### PR TITLE
Paid payment when Opportunity amount is updated from null to a value

### LINKED ISSUE #2763: When Opportunity Amount is updated from null to a value, Payment is not created
[Thread.](https://powerofus.force.com/0D58000003f8sTO)

I'm 99% sure this worked correctly previously, so not sure when a regression was introduced. I know we intentionally had the Payment being created to allow for Opportunity Product line items. 

I replicated the issue in a 3.108 managed scratch org with a Closed/Won Opportunity, but it behaves correctly with an Open one (customer reported having issue with Open opportunity, however).

### PR DESCRIPTION
**Pull request by kofijohnson**
# Critical Changes

# Changes
- NPSP now creates a new Payment, with a Payment Status of Paid, when you update the Amount field on a Closed/Won Opportunity from null or 0 to a value.

# Issues Closed
Fixes #2763
# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by smarovich** [comment]
Hi @kofijohnson, I noticed this when going through my Github notifications. Is this just a bug fix or is there new functionality? If it's just a bug fix, I'd remove the release notes from the Changes section.

**Comment by kofijohnson** [comment]
Hi @smarovich, based on @judisohn 's comment on the ticket (W-020653), I think this is is a new functionality. I will ask again to confirm. Thanks

**Comment by smarovich** [comment]
Thanks, @kofijohnson. I updated the release notes. Let me know if they look ok to you.

File: src/classes/PMT_PaymentCreator.cls [hunk 1]
```diff
@@ -319,29 +319,33 @@ public class PMT_PaymentCreator {
                     }
                     
                     // we may need to add payments if the open opp amount has been changed from $0
```
Comment (joy-sfdo) [inline src/classes/PMT_PaymentCreator.cls]: What are your thoughts on adding the null condition to this comment?

File: src/classes/PMT_PaymentCreator.cls [hunk 2]
```diff
@@ -319,29 +319,33 @@ public class PMT_PaymentCreator {
...
-                    else if (thisOpp.Amount != 0 && thisOpp.Amount != null && (oldOpp.Amount == 0 || oldOpp.Amount == null ) 
-                        && !thisOpp.isClosed && !isNoPaymentForOpp(thisOpp)){
-                        npe01__oppPayment__c op = new npe01__oppPayment__c(
-                            npe01__opportunity__c = thisOpp.id,
-                            npe01__Paid__c = false,
+                    else if (
```
Comment (joy-sfdo) [inline src/classes/PMT_PaymentCreator.cls]: Overall, the code looks. In general, I usually check for null before checking for a value.

File: src/classes/PMT_PaymentCreator.cls [hunk 3]
```diff
@@ -319,29 +319,33 @@ public class PMT_PaymentCreator {
...
+                            thisOpp.Amount != 0 && thisOpp.Amount != null &&
+                            (thisOpp.IsClosed == false || thisOpp.IsWon == true) &&
+                            (oldOpp.Amount == 0 || oldOpp.Amount == null ) &&
```
Comment (joy-sfdo) [inline src/classes/PMT_PaymentCreator.cls]: There is an extra space between null and the closing parenthesis.

File: src/classes/PMT_PaymentCreator_TEST.cls [hunk 4]
```diff
@@ -173,6 +173,115 @@ private with sharing class PMT_PaymentCreator_TEST {
...
+        // create test data
+        Account testAccount = new Account(
+            Name = 'Test_Account'
+        );
+        insert testAccount;
+        Opportunity testOpportunity = new Opportunity(
+            Name = 'TestOpp',
+            AccountId = testAccount.Id,
+            CloseDate = Date.today(), 
+            Amount = null,
```
Comment (joy-sfdo) [inline src/classes/PMT_PaymentCreator_TEST.cls]: What are your thoughts on also testing for $0?

File: src/classes/PMT_PaymentCreator_TEST.cls [hunk 5]
```diff
@@ -173,6 +173,115 @@ private with sharing class PMT_PaymentCreator_TEST {
...
+        // create test data
+        Account testAccount = new Account(
+            Name = 'Test_Account'
+        );
+        insert testAccount;
+        Opportunity testOpportunity = new Opportunity(
+            Name = 'TestOpp',
+            AccountId = testAccount.Id,
+            CloseDate = Date.today(),
+            Amount = null,
```
Comment (joy-sfdo) [inline src/classes/PMT_PaymentCreator_TEST.cls]: Same comment about testing for $0. Thanks.

**Comment by kofijohnson** [comment]
Thank you @smarovich. The release note looks good. I've updated this "from null or 0 to a value". Please let me know if it looks ok. Thanks

**Pull request merged**